sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java

```diff
@@ -1932,6 +1937,141 @@ public TableRow apply(Long input) {
+            .setFields(
+                ImmutableList.of(
+                    new TableFieldSchema().setName("name").setType("STRING"),
+                    new TableFieldSchema().setName("number").setType("INTEGER"),
+                    new TableFieldSchema().setName("tablespec").setType("STRING")));
+    fakeDatasetService.createTable(new Table().setTableReference(tableRef1).setSchema(tableSchema));
+    fakeDatasetService.createTable(new Table().setTableReference(tableRef2).setSchema(tableSchema));
+    fakeDatasetService.createTable(new Table().setTableReference(tableRef3).setSchema(tableSchema));
+
+    LongFunction<Iterable<TableRow>> getRow =
+        (LongFunction<Iterable<TableRow>> & Serializable)
+            (long i) -> {
+              TableRow tableRow;
+              if (i < 5) {
+                tableRow = new TableRow().set("name", "name" + i).set("number", Long.toString(i));
+              } else {
+                tableRow =
+                    new TableRow()
+                        .set("name", "name" + i)
+                        .set("number", Long.toString(i))
+                        .set("double_number", Long.toString(i * 2));
+              }
+              return ImmutableList.of(
+                  tableRow.clone().set("tablespec", tableSpec1),
+                  tableRow.clone().set("tablespec", tableSpec2),
+                  tableRow.clone().set("tablespec", tableSpec3));
+            };
+
+    SerializableFunction<ValueInSingleWindow<TableRow>, TableDestination> tableFunction =
+        value -> new TableDestination((String) value.getValue().get("tablespec"), "");
+
+    final int numRowsPerTable = 1;
+    TestStream.Builder<Long> testStream =
+        TestStream.create(VarLongCoder.of()).advanceWatermarkTo(new Instant(0));
+    for (long i = 0; i < numRowsPerTable; ++i) {
+      testStream = testStream.addElements(i);
+    }
+    testStream = testStream.advanceProcessingTime(Duration.standardSeconds(5));
+
+    PCollection<TableRow> tableRows =
+        p.apply(testStream.advanceWatermarkToInfinity())
+            .apply(
+                MapElements.via(
+                    new SimpleFunction<Long, Iterable<TableRow>>() {
+                      @Override
+                      public Iterable<TableRow> apply(Long input) {
+                        return getRow.apply(input);
+                      }
+                    }))
+            .setCoder(IterableCoder.of(TableRowJsonCoder.of()))
+            .apply(Flatten.iterables());
+    tableRows.apply(
+        BigQueryIO.writeTableRows()
+            .to(tableFunction)
+            .withMethod(method)
+            .withCreateDisposition(BigQueryIO.Write.CreateDisposition.CREATE_NEVER)
+            .withSchemaRefreshFrequency(Duration.standardSeconds(2))
+            .withTestServices(fakeBqServices)
+            .withoutValidation());
+
+    Thread thread =
+        new Thread(
+            () -> {
+              try {
+                Thread.sleep(5000);
```

Comment:
Is there a way where we can reliably advance the schema using teststream and injecting a special element that causes the fakeDatasetService to be updated instead of relying on Thread.sleep?

Reply:
I don't see a way to do it with a special element. DirectRunner blocks on element completion, so we'll never get to the special element. We could add a transform that uses processing-time timers to update the schema, however that's not really any different than using a Thread.